README.md
tyk Cookbook
============

Install and configure Tyk API gateway (https://tyk.io)

## Usage

- For a testing installation, just add the `tyk::default` recipe to your 
runlist. It will install Redis, Mongo, Tyk Gateway, Tyk Dashboard and Tyk Pump
on the node. You have to bootstrap the dashboard with an initial user, as per
manual.
After that, point your browser to your node port 3000 and the dashboard
should come up and let you login with the user you created during the
bootstrap.
- For production, use the provided `install_gateway`, `install_dashboard` and
`install_pump` recipes. You will want to change the configuration attributes.
Best solution is to create your own wrapper cookbook, where you set the
`default` attributes.

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

## License and Authors

Copyright 2016 Internet Mall, a.s.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

### Authors
- Michal Taborsky <michal.taborsky@mall.cz>

### Contributors
- Gaelan D'costa <gdcosta@gmail.com>



CHANGELOG.md
tyk CHANGELOG
=============

This file is used to list changes made in each version of the tyk cookbook.

0.2.0 (2016-05-12)
------------------
- Add rake tests, Rubocop tests currently fail
- Add support for Tyk 2.0 (Thanks to Gaelan D'costa)
- Bump dependencies to latest versions

0.1.2
-----
- Fix configuration attributes for dashboard

0.1.1
-----
- Fix service reload, on some platforms it is not available, so use restart instead. 

0.1.0
-----
- Initial release of tyk cookbook


test/integration/default/serverspec/pump_spec.rb
require 'serverspec'
set :backend, :exec

describe 'tyk-pump' do
  it 'has a service running' do
    expect(service('tyk-pump')).to be_running
  end
end


test/integration/default/serverspec/gateway_spec.rb
require 'serverspec'
set :backend, :exec

describe 'redis' do
  it 'is listening on port 6379' do
    expect(port(6379)).to be_listening
  end
end

describe 'mongodb' do
  it 'is listening on port 27017' do
    expect(port(27_017)).to be_listening
  end
end

describe 'tyk-gateway' do
  it 'has a service running' do
    expect(service('tyk-gateway')).to be_running
  end
  it 'is listening on port 8080' do
    expect(port(8080)).to be_listening
  end
end


test/integration/default/serverspec/dashboard_spec.rb
require 'serverspec'
set :backend, :exec

describe 'tyk-dashboard' do
  it 'has a service running' do
    expect(service('tyk-dashboard')).to be_running
  end
  it 'is listening on port 3000' do
    expect(port(3000)).to be_listening
  end
end


attributes/default.rb
# Gateway
default['tyk']['gateway']['install_path'] = '/opt/tyk-gateway'

# Gateway configuration
# The configuration file is generated from node['tyk']['gateway']['config']
# For documentation see https://tyk.io/docs/tyk-api-gateway-v-2-0/configuration/gateway-configuration-options/
default['tyk']['gateway']['config']['listen_port'] = 8080
default['tyk']['gateway']['config']['secret'] = "Y0uSh0u1dR3a11yChang3Th1sS3cr3t"
default['tyk']['gateway']['config']['node_secret'] = node['tyk']['gateway']['config']['secret']
default['tyk']['gateway']['config']['template_path'] = "#{node['tyk']['gateway']['install_path']}/templates"
default['tyk']['gateway']['config']['tyk_js_path'] = "#{node['tyk']['gateway']['install_path']}/js/tyk.js"
default['tyk']['gateway']['config']['use_db_app_configs'] = false
default['tyk']['gateway']['config']['app_path'] = "#{node['tyk']['gateway']['install_path']}/apps"
default['tyk']['gateway']['config']['middleware_path'] = "#{node['tyk']['gateway']['install_path']}/middleware"
default['tyk']['gateway']['config']['storage']
default['tyk']['gateway']['config']['storage']['type'] = "redis"
default['tyk']['gateway']['config']['storage']['host'] = "localhost"
default['tyk']['gateway']['config']['storage']['port'] = 6379
default['tyk']['gateway']['config']['storage']['username'] = ""
default['tyk']['gateway']['config']['storage']['password'] = ""
default['tyk']['gateway']['config']['storage']['database'] = 0
default['tyk']['gateway']['config']['storage']['optimisation_max_idle'] = 500
default['tyk']['gateway']['config']['storage']['optimisation_max_active'] = 800
default['tyk']['gateway']['config']['enable_analytics'] = false
default['tyk']['gateway']['config']['enable_detailed_recording'] = false
default['tyk']['gateway']['config']['enable_geo_ip'] = false
default['tyk']['gateway']['config']['analytics_config']['ignored_ips'] = []
default['tyk']['gateway']['config']['health_check']
default['tyk']['gateway']['config']['health_check']['enable_health_checks'] = true
default['tyk']['gateway']['config']['health_check']['health_check_value_timeouts'] = 60
default['tyk']['gateway']['config']['optimisations_use_async_session_write'] = true
default['tyk']['gateway']['config']['allow_master_keys'] = false
default['tyk']['gateway']['config']['policies']
default['tyk']['gateway']['config']['policies']['policy_source'] = "file"
default['tyk']['gateway']['config']['policies']['policy_record_name'] = "policies"
default['tyk']['gateway']['config']['hash_keys'] = true
default['tyk']['gateway']['config']['suppress_redis_signal_reload'] = false
default['tyk']['gateway']['config']['close_connections'] = true
default['tyk']['gateway']['config']['local_session_cache']
default['tyk']['gateway']['config']['local_session_cache']['disable_cached_session_state'] = true
default['tyk']['gateway']['config']['uptime_tests']
default['tyk']['gateway']['config']['uptime_tests']['disable'] = false
default['tyk']['gateway']['config']['uptime_tests']['config']
default['tyk']['gateway']['config']['uptime_tests']['config']['enable_uptime_analytics'] = false
default['tyk']['gateway']['config']['uptime_tests']['config']['failure_trigger_sample_size'] = 3
default['tyk']['gateway']['config']['uptime_tests']['config']['time_wait'] = 300
default['tyk']['gateway']['config']['uptime_tests']['config']['checker_pool_size'] = 50
default['tyk']['gateway']['config']['hostname'] = "tyk.local"
default['tyk']['gateway']['config']['enable_custom_domains'] = true
default['tyk']['gateway']['config']['enable_jsvm'] = true

# Dashboard
default['tyk']['dashboard']['install_path'] = '/opt/tyk-dashboard'

# Dashboard configuration
# The configuration file is generated from node['tyk']['dashboard']['config']
# For documentation see https://tyk.io/docs/tyk-dashboard-v1-0/configuration/
default['tyk']['dashboard']['config']['listen_port'] = 3000
default['tyk']['dashboard']['config']['tyk_api_config']['Host'] = "http://localhost"
default['tyk']['dashboard']['config']['tyk_api_config']['Port'] = node['tyk']['gateway']['config']['listen_port'].to_s
default['tyk']['dashboard']['config']['tyk_api_config']['Secret'] = node['tyk']['gateway']['config']['secret']
default['tyk']['dashboard']['config']['mongo_url'] = 'mongodb://127.0.0.1/tyk_analytics'
default['tyk']['dashboard']['config']['shared_node_secret'] = node['tyk']['gateway']['config']['node_secret']
default['tyk']['dashboard']['config']['page_size'] = 10
default['tyk']['dashboard']['config']['admin_secret'] = "12345"
default['tyk']['dashboard']['config']['redis_port'] = node['tyk']['gateway']['config']['storage']['port'] 
default['tyk']['dashboard']['config']['redis_host'] = node['tyk']['gateway']['config']['storage']['host']
default['tyk']['dashboard']['config']['redis_database'] = node['tyk']['gateway']['config']['storage']['database']
default['tyk']['dashboard']['config']['redis_password'] = node['tyk']['gateway']['config']['storage']['password']
default['tyk']['dashboard']['config']['force_api_defaults'] = false
default['tyk']['dashboard']['config']['notify_on_change'] = true
default['tyk']['dashboard']['config']['license_owner'] = "Your Name"
default['tyk']['dashboard']['config']['hash_keys'] = node['tyk']['gateway']['config']['hash_keys'] 
default['tyk']['dashboard']['config']['email_backend'] = {}
default['tyk']['dashboard']['config']['hide_listen_path'] = false
default['tyk']['dashboard']['config']['use_sentry'] = false
default['tyk']['dashboard']['config']['sentry_code'] = ""
default['tyk']['dashboard']['config']['sentry_js_code'] = ""
default['tyk']['dashboard']['config']['show_org_id'] = true
default['tyk']['dashboard']['config']['enable_duplicate_slugs'] = true
default['tyk']['dashboard']['config']['host_config']['override_hostname'] = node['fqdn']
default['tyk']['dashboard']['config']['host_config']['disable_org_slug_prefix'] = true
default['tyk']['dashboard']['config']['host_config']['enable_host_names'] = false
default['tyk']['dashboard']['config']['host_config']['hostname'] = node['fqdn']
default['tyk']['dashboard']['config']['host_config']['portal_domains'] = {}
default['tyk']['dashboard']['config']['host_config']['portal_root_path'] = "/portal"
default['tyk']['dashboard']['config']['http_server_options']['use_ssl'] = false
default['tyk']['dashboard']['config']['http_server_options']['certificates'] = [ {} ]
default['tyk']['dashboard']['config']['ui']['login_page'] = {}
default['tyk']['dashboard']['config']['ui']['nav'] = {}
default['tyk']['dashboard']['config']['ui']['uptime'] = {}
default['tyk']['dashboard']['config']['ui']['portal'] = {}
default['tyk']['dashboard']['config']['ui']['designer'] = {}
default['tyk']['dashboard']['config']['home_dir'] = node['tyk']['dashboard']['install_path']
default['tyk']['dashboard']['config']['identity_broker']['enabled'] = false

# Pump
default['tyk']['pump']['install_path'] = '/opt/tyk-pump'

# Pump configuration
# The configuration file is generated from node['tyk']['pump']['config']
# For documentation see https://tyk.io/docs/tyk-pump/configuration/
default['tyk']['pump']['config']['analytics_storage_type'] = "redis"
default['tyk']['pump']['config']['analytics_storage_config']['type'] = "redis"
default['tyk']['pump']['config']['analytics_storage_config']['host'] = node['tyk']['gateway']['config']['storage']['host']

default['tyk']['pump']['config']['analytics_storage_config']['port'] = node['tyk']['gateway']['config']['storage']['port']

default['tyk']['pump']['config']['analytics_storage_config']['hosts'] = nil
default['tyk']['pump']['config']['analytics_storage_config']['username'] = node['tyk']['gateway']['config']['storage']['username']
default['tyk']['pump']['config']['analytics_storage_config']['password'] = node['tyk']['gateway']['config']['storage']['password']

default['tyk']['pump']['config']['analytics_storage_config']['database'] = node['tyk']['gateway']['config']['storage']['database']

default['tyk']['pump']['config']['analytics_storage_config']['optimisation_max_idle'] = 100
default['tyk']['pump']['config']['analytics_storage_config']['optimisation_max_active'] = 0
default['tyk']['pump']['config']['analytics_storage_config']['enable_cluster'] = false
default['tyk']['pump']['config']['purge_delay'] = 10
default['tyk']['pump']['config']['uptime_pump_config']['collection_name'] = 'tyk_uptime_analytics'
default['tyk']['pump']['config']['uptime_pump_config']['mongo_url'] = node['tyk']['dashboard']['config']['mongo_url']
default['tyk']['pump']['config']['dont_purge_uptime_data'] = false
